CARingBuffer uses CAAudioStreamDescription in a redundant way

kkinnunen-apple opened the pull request:
#### 678b80c4edb8be97cb21b7f1141c9b154c50b591
<pre>
CARingBuffer uses CAAudioStreamDescription in a redundant way
<a href="https://bugs.webkit.org/show_bug.cgi?id=247430">https://bugs.webkit.org/show_bug.cgi?id=247430</a>
rdar://problem/101914553

Reviewed by Eric Carlson.

CARingBuffer structure is defined by channelCount, bytesPerFrame, frameCount. However,
it also stores CAAudioStreamDescription to define the contents of each frame
in case client wants to mix (contents can be floats, ints of different size).
CARingBuffer might also be used to transfer undefined frames (possibly non-PCM data), via the
copy operation.

Using CAAudioStreamDescription is problematic, as it is redundant information and it smuggles
the non-PCM data definition via the platform-specific descriptor which is not consistent.
In order to stop using CAAudioStreamDescription in the future patches, remove it from
CARingBuffer. The fetch() caller already knows what kind of data is expected, so this information
is redundant. This fixes the problem that the caller expectation and what is inside the
CARingBuffer could diverge.

Leave the use of CAAudioStreamDescription in the constructors, but remove it from the
internals of the classes.

This patch would affect the base-class constructors and functions. Take the opportunity
to change the CARingBuffer allocation logic: previously, one would construct an empty
instance and then at some point call allocate(). In case the allocate would fail, the
instance would stay &quot;empty&quot; with quite undefined behavior. Instead, make all CARingBuffer
classes be normal memory buffer classes: creation will allocate the buffer, and this
might fail. Deletion will deallocate.

Previously ProducerSharedCARingBuffer allocation sequence was a bit convoluted:
- WP proxy class would construct empty instance with a resize callback.
- When allocate and deallocate would succeed, the resize callback would be called.
- The callers would construct with a resize callback that would then
  send the &quot;audio storage changed&quot; message to the other process, typically GPUP.
The sequence would be linear, resize callback invocation would always follow
an allocate. However, it was written in non-linear fashion.

Instead just have a linear logic:
- WP proxy class constructs the shared buffer, and gets the buffer and
  client handle as a result.
- WP proxy class sends the client handle to the other process.

In order to return ConsumerSharedCARingBuffer::Handle (alias for
SharedMemory::Handle) as part of a struct, make it copyable.
Recent changes to SharedMemory::Handle enable this and were made
to for this purpose.

* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.h:
* Source/WebCore/platform/audio/cocoa/AudioSampleDataSource.h:
* Source/WebCore/platform/audio/cocoa/AudioSampleDataSource.mm:
(WebCore::AudioSampleDataSource::setOutputFormat):
(WebCore::AudioSampleDataSource::pushSamplesInternal):
(WebCore::AudioSampleDataSource::pullSamples):
(WebCore::AudioSampleDataSource::pullSamplesInternal):
(WebCore::AudioSampleDataSource::pullAvailableSampleChunk):
(WebCore::AudioSampleDataSource::pullAvailableSamplesAsChunks):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::CARingBuffer::CARingBuffer):
(WebCore::CARingBuffer::computeCapacityBytes):
(WebCore::CARingBuffer::computeSizeForBuffers):
(WebCore::CARingBuffer::initialize):
(WebCore::FetchABL):
(WebCore::CARingBuffer::fetchInternal):
(WebCore::InProcessCARingBuffer::allocate):
(WebCore::InProcessCARingBuffer::InProcessCARingBuffer):
(WebCore::CARingBuffer::initializeAfterAllocation): Deleted.
(WebCore::CARingBuffer::allocate): Deleted.
(WebCore::CARingBuffer::deallocate): Deleted.
(WebCore::InProcessCARingBuffer::allocateBuffers): Deleted.
* Source/WebCore/platform/audio/cocoa/CARingBuffer.h:
(WebCore::CARingBuffer::fetchModeForMixing):
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.mm:
(WebCore::AudioSourceProviderAVFObjC::AudioSourceProviderAVFObjC):
(WebCore::AudioSourceProviderAVFObjC::prepare):
(WebCore::AudioSourceProviderAVFObjC::process):
(WebCore::AudioSourceProviderAVFObjC::setConfigureAudioStorageCallback):
(WebCore::AudioSourceProviderAVFObjC::setRingBufferCreationCallback): Deleted.
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
* Source/WebKit/GPUProcess/media/RemoteAudioSourceProviderProxy.cpp:
(WebKit::RemoteAudioSourceProviderProxy::create):
(WebKit::RemoteAudioSourceProviderProxy::configureAudioStorage):
(WebKit::RemoteAudioSourceProviderProxy::createRingBuffer): Deleted.
(WebKit::RemoteAudioSourceProviderProxy::storageChanged): Deleted.
* Source/WebKit/GPUProcess/media/RemoteAudioSourceProviderProxy.h:
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit::start):
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.cpp:
(WebKit::RemoteMediaRecorder::audioSamplesStorageChanged):
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
* Source/WebKit/Platform/unix/SharedMemoryUnix.cpp:
(WebKit::SharedMemory::Handle::Handle): Deleted.
(WebKit::SharedMemory::Handle::~Handle): Deleted.
* Source/WebKit/Platform/win/SharedMemoryWin.cpp:
* Source/WebKit/Shared/Cocoa/SharedCARingBuffer.cpp:
(WebKit::SharedCARingBufferBase::SharedCARingBufferBase):
(WebKit::SharedCARingBufferBase::data):
(WebKit::SharedCARingBufferBase::sharedFrameBounds const):
(WebKit::SharedCARingBufferBase::updateFrameBounds):
(WebKit::SharedCARingBufferBase::size const):
(WebKit::ConsumerSharedCARingBuffer::map):
(WebKit::ProducerSharedCARingBuffer::allocate):
(WebKit::ProducerSharedCARingBuffer::setCurrentFrameBounds):
(WebKit::ConsumerSharedCARingBuffer::ConsumerSharedCARingBuffer): Deleted.
(WebKit::ConsumerSharedCARingBuffer::allocateBuffers): Deleted.
(WebKit::ProducerSharedCARingBuffer::setStorage): Deleted.
(WebKit::ProducerSharedCARingBuffer::allocateBuffers): Deleted.
(WebKit::ProducerSharedCARingBuffer::deallocateBuffers): Deleted.
* Source/WebKit/Shared/Cocoa/SharedCARingBuffer.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::~SourceProxy):
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::storageChanged): Deleted.
* Source/WebKit/UIProcess/SpeechRecognitionRemoteRealtimeMediaSource.cpp:
(WebKit::SpeechRecognitionRemoteRealtimeMediaSource::setStorage):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::RemoteAudioDestinationProxy):
(WebKit::RemoteAudioDestinationProxy::connection):
(WebKit::RemoteAudioDestinationProxy::storageChanged): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProviderManager.cpp:
(WebKit::RemoteAudioSourceProviderManager::RemoteAudio::setStorage):
* Source/WebKit/WebProcess/GPU/webrtc/AudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::start):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::storageChanged): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.cpp:
(WebKit::MediaRecorderPrivate::startRecording):
(WebKit::MediaRecorderPrivate::audioSamplesAvailable):
(WebKit::MediaRecorderPrivate::storageChanged): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.h:
* Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::Source::Source):
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::Source::~Source):
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::Source::storageChanged): Deleted.
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.cpp:
(WebKit::RemoteCaptureSampleManager::RemoteAudio::setStorage):
* Tools/TestWebKitAPI/Tests/WebCore/CARingBuffer.cpp:
(TestWebKitAPI::CARingBufferTest::SetUp):
(TestWebKitAPI::CARingBufferTest::setup):
(TestWebKitAPI::MixingTest::run):

Canonical link: <a href="https://commits.webkit.org/256392@main">https://commits.webkit.org/256392@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bcdbbc3bdef76e90c8235295856fc504d5ef0e0d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95651 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4912 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28699 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105229 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165529 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99636 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4979 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33667 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88025 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101078 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101312 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82269 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/30711 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87428 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/73543 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39401 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37100 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20282 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4419 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41096 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43083 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39533 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/GPUProcess/media/RemoteAudioSourceProviderProxy.cpp:
```diff
@@ -55,26 +55,20 @@ RemoteAudioSourceProviderProxy::RemoteAudioSourceProviderProxy(WebCore::MediaPla
-    });
+    auto [ringBuffer, handle] = ProducerSharedCARingBuffer::allocate(format, frameCount);
+    m_connection->send(Messages::RemoteAudioSourceProviderManager::AudioStorageChanged { m_identifier, WTFMove(handle), format, frameCount }, 0);
+    // Use a redundant variable to avoid move in return position and to obtain copy elision. Clang or libc++ does not allow returning covariant of Ts from std::unique_ptr<T>s in this position.å
```
Nit: is the `å` just a git artifact or is there a rogue \u00e5 in the comment?

eric-carlson reviewed (approved):
Very nice!

kkinnunen-apple committed:
CARingBuffer uses CAAudioStreamDescription in a redundant way
https://bugs.webkit.org/show_bug.cgi?id=247430
rdar://problem/101914553

Reviewed by Eric Carlson.

CARingBuffer structure is defined by channelCount, bytesPerFrame, frameCount. However,
it also stores CAAudioStreamDescription to define the contents of each frame
in case client wants to mix (contents can be floats, ints of different size).
CARingBuffer might also be used to transfer undefined frames (possibly non-PCM data), via the
copy operation.

Using CAAudioStreamDescription is problematic, as it is redundant information and it smuggles
the non-PCM data definition via the platform-specific descriptor which is not consistent.
In order to stop using CAAudioStreamDescription in the future patches, remove it from
CARingBuffer. The fetch() caller already knows what kind of data is expected, so this information
is redundant. This fixes the problem that the caller expectation and what is inside the
CARingBuffer could diverge.

Leave the use of CAAudioStreamDescription in the constructors, but remove it from the
internals of the classes.

This patch would affect the base-class constructors and functions. Take the opportunity
to change the CARingBuffer allocation logic: previously, one would construct an empty
instance and then at some point call allocate(). In case the allocate would fail, the
instance would stay "empty" with quite undefined behavior. Instead, make all CARingBuffer
classes be normal memory buffer classes: creation will allocate the buffer, and this
might fail. Deletion will deallocate.

Previously ProducerSharedCARingBuffer allocation sequence was a bit convoluted:
- WP proxy class would construct empty instance with a resize callback.
- When allocate and deallocate would succeed, the resize callback would be called.
- The callers would construct with a resize callback that would then
  send the "audio storage changed" message to the other process, typically GPUP.
The sequence would be linear, resize callback invocation would always follow
an allocate. However, it was written in non-linear fashion.

Instead just have a linear logic:
- WP proxy class constructs the shared buffer, and gets the buffer and
  client handle as a result.
- WP proxy class sends the client handle to the other process.

In order to return ConsumerSharedCARingBuffer::Handle (alias for
SharedMemory::Handle) as part of a struct, make it copyable.
Recent changes to SharedMemory::Handle enable this and were made
to for this purpose.

* Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.h:
* Source/WebCore/platform/audio/cocoa/AudioSampleDataSource.h:
* Source/WebCore/platform/audio/cocoa/AudioSampleDataSource.mm:
(WebCore::AudioSampleDataSource::setOutputFormat):
(WebCore::AudioSampleDataSource::pushSamplesInternal):
(WebCore::AudioSampleDataSource::pullSamples):
(WebCore::AudioSampleDataSource::pullSamplesInternal):
(WebCore::AudioSampleDataSource::pullAvailableSampleChunk):
(WebCore::AudioSampleDataSource::pullAvailableSamplesAsChunks):
* Source/WebCore/platform/audio/cocoa/CARingBuffer.cpp:
(WebCore::CARingBuffer::CARingBuffer):
(WebCore::CARingBuffer::computeCapacityBytes):
(WebCore::CARingBuffer::computeSizeForBuffers):
(WebCore::CARingBuffer::initialize):
(WebCore::FetchABL):
(WebCore::CARingBuffer::fetchInternal):
(WebCore::InProcessCARingBuffer::allocate):
(WebCore::InProcessCARingBuffer::InProcessCARingBuffer):
(WebCore::CARingBuffer::initializeAfterAllocation): Deleted.
(WebCore::CARingBuffer::allocate): Deleted.
(WebCore::CARingBuffer::deallocate): Deleted.
(WebCore::InProcessCARingBuffer::allocateBuffers): Deleted.
* Source/WebCore/platform/audio/cocoa/CARingBuffer.h:
(WebCore::CARingBuffer::fetchModeForMixing):
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/AudioSourceProviderAVFObjC.mm:
(WebCore::AudioSourceProviderAVFObjC::AudioSourceProviderAVFObjC):
(WebCore::AudioSourceProviderAVFObjC::prepare):
(WebCore::AudioSourceProviderAVFObjC::process):
(WebCore::AudioSourceProviderAVFObjC::setConfigureAudioStorageCallback):
(WebCore::AudioSourceProviderAVFObjC::setRingBufferCreationCallback): Deleted.
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
* Source/WebKit/GPUProcess/media/RemoteAudioSourceProviderProxy.cpp:
(WebKit::RemoteAudioSourceProviderProxy::create):
(WebKit::RemoteAudioSourceProviderProxy::configureAudioStorage):
(WebKit::RemoteAudioSourceProviderProxy::createRingBuffer): Deleted.
(WebKit::RemoteAudioSourceProviderProxy::storageChanged): Deleted.
* Source/WebKit/GPUProcess/media/RemoteAudioSourceProviderProxy.h:
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit::start):
* Source/WebKit/GPUProcess/webrtc/RemoteMediaRecorder.cpp:
(WebKit::RemoteMediaRecorder::audioSamplesStorageChanged):
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
* Source/WebKit/Platform/unix/SharedMemoryUnix.cpp:
(WebKit::SharedMemory::Handle::Handle): Deleted.
(WebKit::SharedMemory::Handle::~Handle): Deleted.
* Source/WebKit/Platform/win/SharedMemoryWin.cpp:
* Source/WebKit/Shared/Cocoa/SharedCARingBuffer.cpp:
(WebKit::SharedCARingBufferBase::SharedCARingBufferBase):
(WebKit::SharedCARingBufferBase::data):
(WebKit::SharedCARingBufferBase::sharedFrameBounds const):
(WebKit::SharedCARingBufferBase::updateFrameBounds):
(WebKit::SharedCARingBufferBase::size const):
(WebKit::ConsumerSharedCARingBuffer::map):
(WebKit::ProducerSharedCARingBuffer::allocate):
(WebKit::ProducerSharedCARingBuffer::setCurrentFrameBounds):
(WebKit::ConsumerSharedCARingBuffer::ConsumerSharedCARingBuffer): Deleted.
(WebKit::ConsumerSharedCARingBuffer::allocateBuffers): Deleted.
(WebKit::ProducerSharedCARingBuffer::setStorage): Deleted.
(WebKit::ProducerSharedCARingBuffer::allocateBuffers): Deleted.
(WebKit::ProducerSharedCARingBuffer::deallocateBuffers): Deleted.
* Source/WebKit/Shared/Cocoa/SharedCARingBuffer.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::~SourceProxy):
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::storageChanged): Deleted.
* Source/WebKit/UIProcess/SpeechRecognitionRemoteRealtimeMediaSource.cpp:
(WebKit::SpeechRecognitionRemoteRealtimeMediaSource::setStorage):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::RemoteAudioDestinationProxy):
(WebKit::RemoteAudioDestinationProxy::connection):
(WebKit::RemoteAudioDestinationProxy::storageChanged): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSourceProviderManager.cpp:
(WebKit::RemoteAudioSourceProviderManager::RemoteAudio::setStorage):
* Source/WebKit/WebProcess/GPU/webrtc/AudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::start):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::storageChanged): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.cpp:
(WebKit::MediaRecorderPrivate::startRecording):
(WebKit::MediaRecorderPrivate::audioSamplesAvailable):
(WebKit::MediaRecorderPrivate::storageChanged): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/MediaRecorderPrivate.h:
* Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::Source::Source):
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::Source::~Source):
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::Source::storageChanged): Deleted.
* Source/WebKit/WebProcess/cocoa/RemoteCaptureSampleManager.cpp:
(WebKit::RemoteCaptureSampleManager::RemoteAudio::setStorage):
* Tools/TestWebKitAPI/Tests/WebCore/CARingBuffer.cpp:
(TestWebKitAPI::CARingBufferTest::SetUp):
(TestWebKitAPI::CARingBufferTest::setup):
(TestWebKitAPI::MixingTest::run):

Canonical link: https://commits.webkit.org/256392@main
